[EDGE-1520] Add pipeline templates rest & graphql api documentation (#2479)
buildkite/docs: pages/apis/graphql/cookbooks/pipeline_templates.md

    name: "template name",
    description: "it does a thing",
    configuration: "steps:\n  - command: deploy.sh",
    available: false
  }) {
    pipelineTemplate {
      id
      uuid
      name
      description
      configuration
      available
    }
  }
}
```

## Updating a pipeline template

Update a pipeline template on an organization using the `pipelineTemplateUpdate` mutation, specifying the ID for organization and pipeline template:

```graphql
mutation UpdatePipelineTemplate {
  pipelineTemplateUpdate(input: {
    organizationId: "organization-id",
    id: "pipeline-template-id",
    configuration: "steps:\n - comand: updated_steps.sh"
    available: true
  }) {
    pipelineTemplate {
      id
      uuid
      name
      description
      configuration
      available
    }
  }
}
```

## Deleting a pipeline template

Delete a pipeline template using the `pipelineTemplateDelete` mutation, specifying the ID for organization and pipeline template:

```graphql
mutation DeletePipelineTemplate {
  pipelineTemplateDelete(input: {
    organizationId: "organization-id",
    id: "pipeline-template-id"
  }) {
    deletedPipelineTemplateId
  }
}
```

> ## Managing pipeline template assignment on a pipeline

Reviewer: ```suggestion
## Manage pipeline template assignment on a pipeline
```

Reviewer: Actually, I wonder if you should split out the two actions rather than grouping them together. Thinking:

```md
## Assign a template to a pipeline

...

## Remove a template from a pipeline

...
```

I think that's probably clearer and would help people find the task in the table of contents and search. 🤔

Author: Great suggestion @mbelton-buildkite. I tossed between the two ideas too but didn't think about the contents sidebar

> Admins and users with manage pipeline permissions can assign a pipeline template to a pipeline using the `pipelineUpdate` mutation:

Reviewer: Is "manage pipeline" the name of the permission in the UI? If so, I think it would be good to use italics:

```suggestion
Admins and users with _manage pipeline permissions_ can assign a pipeline template to a pipeline using the `pipelineUpdate` mutation:
```

Or, consider the following alternate, which I think reads a little better:

```suggestion
Admins and users with permission to manage pipelines can assign a pipeline template to a pipeline using the `pipelineUpdate` mutation:
```

Author: "Manage pipeline" isn't the permission we use in the UI, I used it to collectively describe `read_project` and `create_project` permissions :).

Do you think we should be more specific with the permissions?

Reviewer: Ah okay. Hmmm is there anywhere else that tells them that specific info? The first place I checked was our GraphQL schema docs, but that doesn't include permission requirements. The next place I looked was the Pipeline templates page, which has:

> anyone with permission to create or change a pipeline can assign a template.

Is that the equivalent info?

If so, I think it's fine to just use my second suggestion: "Admins and users with permission to manage pipelines can..."

Author: The closest page I can think of that has a similar granularity on permissions would be on the pipeline templates rest api page. Each of the endpoints list out what permissions and api token scopes are required in order to use the api.

Cool, I ran with your second suggestion in lieu of any other updates we may want to make in the future :)

> ## Managing pipeline template assignment on a pipeline
> 
> Admins and users with manage pipeline permissions can assign a pipeline template to a pipeline using the `pipelineUpdate` mutation:
> 
> ```graphql
> mutation AssignPipelineTemplate {
>   pipelineUpdate(input: {
>     id: 'pipeline-id'
>     pipelineTemplateId: 'pipeline-template-id'
>   }) {
>     pipeline {
>       id
>       name
>       pipelineTemplateId
>     }
>   }
> }
> ```
> 
> Conversely, pipeline templates can be removed from a pipeline by specifying `pipelineTemplateId` as `null` in the mutation input:
> 
> ```graphql
> mutation UnassignPipelineTemplate {
>   pipelineUpdate(input: {
>     id: 'pipeline-id'
>     pipelineTemplateId: null
>   }) {
>     pipeline {
>       id
>       name
>       pipelineTemplateId
>     }
>   }
> }
> ```

Reviewer 2: Seems like this should be using double quotes (syntax error is highlighted in the generated docs too):

<img width="681" alt="image" src="https://github.com/buildkite/docs/assets/673828/dacdb34b-5c4b-4d76-aa0c-6061b940ed71">

````suggestion
## Managing pipeline template assignment on a pipeline

Admins and users with manage pipeline permissions can assign a pipeline template to a pipeline using the `pipelineUpdate` mutation:

```graphql
mutation AssignPipelineTemplate {
  pipelineUpdate(input: {
    id: "pipeline-id"
    pipelineTemplateId: "pipeline-template-id"
  }) {
    pipeline {
      id
      name
      pipelineTemplateId
    }
  }
}
```

Conversely, pipeline templates can be removed from a pipeline by specifying `pipelineTemplateId` as `null` in the mutation input:

```graphql
mutation UnassignPipelineTemplate {
  pipelineUpdate(input: {
    id: "pipeline-id"
    pipelineTemplateId: null
  }) {
    pipeline {
      id
      name
      pipelineTemplateId
    }
  }
}
```
````

Author: Ah great pick up @mhaylock, fixed up the syntax in [8f472e4](https://github.com/buildkite/docs/pull/2479/commits/8f472e4e12307e1ba5011e4eb4d0c9adb608d6d2)

```graphql
mutation UpdatePipelineTemplate {
  pipelineTemplateUpdate(input: {
    organizationId: "organization-id",
    id: "pipeline-template-id",
    configuration: "steps:\n - comand: updated_steps.sh"
    available: true
  }) {
    pipelineTemplate {
      id
      uuid
      name
      description
      configuration
      available
    }
  }
}
```

## Delete a pipeline template

Delete a pipeline template using the `pipelineTemplateDelete` mutation, specifying the ID for organization and pipeline template:

```graphql
mutation DeletePipelineTemplate {
  pipelineTemplateDelete(input: {
    organizationId: "organization-id",
    id: "pipeline-template-id"
  }) {
    deletedPipelineTemplateId
  }
}
```

## Assign a template to a pipeline

Admins and users with permission to manage pipelines can assign a pipeline template to a pipeline using the `pipelineUpdate` mutation:

```graphql
mutation AssignPipelineTemplate {
  pipelineUpdate(input: {
    id: "pipeline-id"
    pipelineTemplateId: "pipeline-template-id"
  }) {
    pipeline {
      id
      name
      pipelineTemplateId
    }
  }
}
```

## Remove a template from a pipeline

> Admins and users with permission to manage pipeliness can remove from a pipeline by specifying `pipelineTemplateId` as `null` in the mutation input:

Reviewer: Typo:

```suggestion
Admins and users with permission to manage pipelines can remove from a pipeline by specifying `pipelineTemplateId` as `null` in the mutation input:
```

(this should fix the failing build)

Reviewer: Looks like it's also missing "a pipeline template," too:

```suggestion
Admins and users with permission to manage pipelines can remove a pipeline template from a pipeline by specifying `pipelineTemplateId` as `null` in the mutation input:
```
